### PR TITLE
[SPARK-23408][SS][BRANCH-2.3] Synchronize successive AddData actions in Streaming*JoinSuite

### PR DESCRIPTION
**Pull request by HeartSaVioR**
## What changes were proposed in this pull request?

**The best way to review this PR is to ignore whitespace/indent changes. Use this link - https://github.com/apache/spark/pull/20650/files?w=1**

The stream-stream join tests add data to multiple sources and expect it all to show up in the next batch. But there's a race condition; the new batch might trigger when only one of the AddData actions has been reached.

Prior attempt to solve this issue by jose-torres in #20646 attempted to simultaneously synchronize on all memory sources together when consecutive AddData was found in the actions. However, this carries the risk of deadlock as well as unintended modification of stress tests (see the above PR for a detailed explanation). Instead, this PR attempts the following.

- A new action called `StreamProgressBlockedActions` that allows multiple actions to be executed while the streaming query is blocked from making progress. This allows data to be added to multiple sources that are made visible simultaneously in the next batch.
- An alias of `StreamProgressBlockedActions` called `MultiAddData` is explicitly used in the `Streaming*JoinSuites` to add data to two memory sources simultaneously.

This should avoid unintentional modification of the stress tests (or any other test for that matter) while making sure that the flaky tests are deterministic.

NOTE: This patch is modified a bit from origin PR (#20650) to cover DSv2 incompatibility between Spark 2.3 and 2.4: StreamingDataSourceV2Relation is a class for 2.3, whereas it is a case class for 2.4

## How was this patch tested?

Modified test cases in `Streaming*JoinSuites` where there are consecutive `AddData` actions.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
NOTE: e08bf2e (my own commit) should be flattened into be41fa7 - I'll sort out later. I would like to confirm flakiness is resolved first.

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HeartSaVioR** [comment]
I wish d5afd92 would be the last one needed for resolving flakiness tests. Only one failed for e08bf2e and the test failure is related to d5afd92.

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
cc. @maropu @srowen 
This PR is a set of commits for porting back so IMO we can just pull the branch to apply and close this manually: please let me know when we would like to squash into one - I'll update PR's title and description.

**Comment by HeartSaVioR** [comment]
More clearly, this PR intends to address https://issues.apache.org/jira/browse/SPARK-24211 but while addressing it triggers https://issues.apache.org/jira/browse/SPARK-24239 (may want to change affect version?) so also the fix of SPARK-24239 is included to the PR. My understanding is we could resolve them with setting version to 2.3.4.

**Comment by srowen** [comment]
BTW is this a 'port' of a similar fix from later branches? it would be good to cross-reference those.
You can add the JIRAs that this addresses to the title, unless we're not sure this fully fixes it.

**Comment by HeartSaVioR** [comment]
Yes, now this patch is just a set of commits ported back from higher branch, with some modification to fill the gap between Spark 2.3 and 2.4 (DSv2 changed a bit).

I just added JIRA issues which refer test flakiness to PR title. If you meant we would be better to add origin JIRA issues (actual commits) to the title, please let me know.

**Comment by HeartSaVioR** [comment]
> **[Test build #102161 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102161/testReport)** for PR 23757 at commit [`d9fcb66`](https://github.com/apache/spark/commit/d9fcb66c9da66546a632d06ae35d06f998305306).
> 
> * This patch **fails Spark unit tests**.
> * This patch merges cleanly.
> * This patch adds no public classes.

This failure seems not relevant to the change: other builds passed.

```
Failed
org.apache.spark.sql.FileBasedDataSourceSuite.(It is not a test it is a sbt.testing.SuiteSelector)

Error Message
org.scalatest.exceptions.TestFailedDueToTimeoutException: The code passed to eventually never returned normally. Attempted 15 times over 10.00938722 seconds. Last failure message: There are 1 possibly leaked file streams..
```

**Comment by HeartSaVioR** [comment]
I'll run four more set of build triggers - it would be 10 times including state before rebasing.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by srowen** [comment]
Just for reference, what are the other commits you're looking at, for comparison? that will help review. It's OK if they need to be modified a bit to back-port but just want to make sure it's not making additional (maybe helpful) changes just in the back port.

**Comment by HeartSaVioR** [comment]
Ah OK. That's great idea. Here're pairs of origin and modified commits for cherry-picking.

JIRA issue | origin | cherry-picked
-----------|--------|---------------
[SPARK-23408](https://issues.apache.org/jira/browse/SPARK-23408) | 855ce13d045569b7b16fdc7eee9c981f4ff3a545 | 9fa380f
[SPARK-23491](https://issues.apache.org/jira/browse/SPARK-23491) | 7ec8365 | d9fcb66
[SPARK-23416](https://issues.apache.org/jira/browse/SPARK-23416) | f457933 | 8bd0f1f

Will update to PR's description as well.

**Comment by HeartSaVioR** [comment]
NOTE: looks like three pairs are identical - there's something got mixed up when cherry-picking and made me think I have fixed some incompatibility. I'll sort out commit description.

Please let me know if we would also want to not introduce minor change (removed unused imports).

**Comment by HeartSaVioR** [comment]
New failures (3 of 4 of builds) don't look related to streaming join as well as kafka continuous mode.

The failures look relevant to [SPARK-23390](https://issues.apache.org/jira/browse/SPARK-23390) and relevant issues. The major fix [SPARK-26427](https://issues.apache.org/jira/browse/SPARK-26427) looks requiring ORC version upgrade, so not sure we can take it forward for branch-2.3.

**Comment by maropu** [comment]
Thanks for your great help, @HeartSaVioR! Just to make sure, I have one question; we need `SPARK-23408` for fixing the streaming outer join failures, but `SPARK-23408` invokes the other Kafka test failures. To fix it, we need more backports, `SPARK-23491` and `SPARK-23416`, right?

Also, I just want to check the diff of the original and your fix in `SPARK-23408`.

**Comment by HeartSaVioR** [comment]
I'm finalizing the PR via minimizing the change between origin and cherry-picked. I was right: there was incompatibility on DSv2 between Spark 2.3 and Spark 2.4 so I had to fix them. Will update the PR description once I sort out and rebase.

Regarding relevant of SPARK-23408 and SPARK-23491 / SPARK-23416, now I'm not 100% sure of relevant between them, because I just realized the test failure can hide other test failures which are not relevant. If we really want to split out PRs via relevance, we might be able to have two individual PRs, but need more times to split them and test them individually, so not sure about needs it cause this is just a matter of porting back.

**Comment by maropu** [comment]
If we merge the 3 commits in branch-2.3, I think we merge the `SPARK-23408` commit with your fix in this pr first (This commit has conflicts in branch-2.3, right?). And then, we'll just cherry-pick the other two commits (SPARK-23491/SPARK-23416) later if they have no conflict.

But, I'm not familiar with this part, so we need strict final checks before merging them.

**Comment by HeartSaVioR** [comment]
If Spark project allows pulling commits from some branch, just pulling commits (merge branch with fast forward) would work smoothly. But I'm not sure whether there's some policies related to this. At least I guess merge script wouldn't work for this PR.

**Comment by maropu** [comment]
> If Spark project allows pulling commits from some branch, just pulling commits (merge branch with fast forward) would work smoothly (cause I already fixed the issue per commit during cherry-picking).

Yea, we can do. But, I checked on my laptop; I could cleanly cherry-pick SPARK-23491/SPARK-23416 in branch-2.3 though, I couldn't do SPARK-23408. So, we need to make a new pr in branch-2.3 based on the SPARK-23408 commit (probably, this pr?), and then review/merge it in branch-2.3.

Anyway, we need to check its ok to merge the 3 commits in branch-2.3 first.

**Comment by HeartSaVioR** [comment]
What I have in mind is like:

```
git checkout branch-2.3
git checkout -b branch-2.3-merging-23757
git pull https://github.com/HeartSaVioR/spark.git fix-streaming-join-test-flakiness-branch-2.3
git log
# should contain a6662f27208622cdd63b2ce48fb336d14a488f17 / 079a83bd33cf38ed8a4ba1e5b41bc4c654cbf35c / ddd68fba8840f0191f9820aab2911390ace47c8d
git checkout branch-2.3
git cherry-pick a6662f27208622cdd63b2ce48fb336d14a488f17
# cherry-pick either origin commits (7ec83658fbc88505dfc2d8a6f76e90db747f1292 / f4579332931c9bf424d0b6147fad89bd63da26f6) or cherry-picked commits (079a83bd33cf38ed8a4ba1e5b41bc4c654cbf35c / ddd68fba8840f0191f9820aab2911390ace47c8d)
```

I'm not sure what you meant, but if you mean we have to have explicit PR for cherry-picked commit (a6662f27208622cdd63b2ce48fb336d14a488f17) with modification, I'll just remove two commits from here, and sync PR's title and description to SPARK-23408 so that merge script would work. After that you could cherry pick origin commits for SPARK-23491/SPARK-23416. Do I understand correctly?

**Comment by maropu** [comment]
Yea, yes. I meant so (I'm sorry ambiguous).

**Comment by HeartSaVioR** [comment]
Just removed two commits and modified PR's title and description.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Test failed from `org.apache.spark.sql.FileBasedDataSourceSuite.(It is not a test it is a sbt.testing.SuiteSelector)`

Retest this, please

**Comment by HeartSaVioR** [comment]
Retest this, please

**Comment by srowen** [comment]
Just to make sure I understand the status:

This is a straight back-port of https://github.com/apache/spark/pull/20650/files?w=1 , and indeed it looks like the same change we see here with https://github.com/apache/spark/pull/23757/files?w=1 , OK. It can be merged first without other changes, pending tests?

Maybe someone can help me understand this. The diff at 
https://github.com/apache/spark/pull/23757/files?w=1
looks pretty different from
https://github.com/apache/spark/pull/23757/files

I know the difference is supposed to be just whitespace in the diff, but the latter seems to show a much more significant reordering of code. I can't figure out what to make of it. Am I missing something obvious?

**Comment by HeartSaVioR** [comment]
> This is a straight back-port of https://github.com/apache/spark/pull/20650/files?w=1 , and indeed it looks like the same change we see here with https://github.com/apache/spark/pull/23757/files?w=1 , OK. It can be merged first without other changes, pending tests?

Yes (and you can also resolve SPARK-24211 with this patch). And then SPARK-23491 and SPARK-23416 can be cherry-picked from origin PRs since they're clean cherry-pick and we already ran tests in this PR before revising.(and you can also resolve SPARK-24239 with cherry-picking them,)

> Maybe someone can help me understand this. The diff at
> https://github.com/apache/spark/pull/23757/files?w=1
> looks pretty different from
> https://github.com/apache/spark/pull/23757/files

It is just same as what #20650 was. You could check the difference between https://github.com/apache/spark/pull/23757/files?w=1 vs https://github.com/apache/spark/pull/23757/files

The code change add indentation(s) to huge code block: `w=1` option ignores differences of indentation(s) whereas no option takes them all diff. You could find the difference from searching `logInfo(s"Processing test stream action: $action")` from PR with and without option.

BTW, after fixing this I'm seeing very high change of failures from FileBasedDataSourceSuite - as I commented earlier it might need ORC version upgrade to fix some of related issues, so while we seemed to upgrade the ORC dependency from bugfix version (2.4.0 -> 2.4.1) I'm not sure we could do in 2.3 version line.
cc. @dongjoon-hyun Could you help explaining how we dealt with FileBasedDataSourceSuite and relevant test failures? I'm not sure I understand all of changes.

**Comment by HeartSaVioR** [comment]
IMHO once we know we're not adding code to branch-2.3, passed builds could be a complement of pending test. If we need to run test just before merging, I'd propose triggering 5 builds concurrently due to high chance of failure of `FileBasedDataSourceSuite`.

**Comment by srowen** [comment]
OK after staring at it longer, I see that the difference between the diffs is actually whitespace. It looks much bigger because it's mismatching lines that occur several time identically in different parts of the code, so it renders here as a big delete and add. OK disregard that, just wanted to make sure I'm not crazy.

This is looking good pending tests. Once this passes and is merged, go ahead with any other changes as you see fit.

**Comment by HeartSaVioR** [comment]
I'm not a committer so I need committer's help on cherry-picking remaining two commits (SPARK-23491 and SPARK-23416). Not sure we would want to have PRs for clean cherry-pick: if we would like to have them explicitly please let me know - I'll submit PRs.

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Let me run three builds concurrently: as build history from this PR, there's other flaky test which have high chance of failure as well.

**Comment by maropu** [comment]
Is the single line below different from the original one?
https://github.com/apache/spark/pull/23757/files?w=1#diff-77e3d95970339f518398191d11e3bb8dR652

cc: @dongjoon-hyun could you help?

**Comment by HeartSaVioR** [comment]
Yes right. Maybe possibly other line but it will be just an import.

**Comment by maropu** [comment]
ok, thanks! Looks ok to me.
(Probably, for reviews, you'd better to separate a modified commit from an original one next time)

**Review by srowen — APPROVED** [review]
Looks good to me. Just so I'm 100% clear, this is intended to be merged first right?

**Comment by HeartSaVioR** [comment]
Yes right.

**Comment by srowen** [comment]
Merged to 2.3.

**Comment by HeartSaVioR** [comment]
Thanks for merging!

As a follow-up, we may want to consider doing these things to resolve flaky tests specifically in branch-2.3:

1.  port back [SPARK-23491](https://issues.apache.org/jira/browse/SPARK-23491) and [SPARK-23416](https://issues.apache.org/jira/browse/SPARK-23416)  (don't seem to need to have PRs)
2. discuss regarding porting back [SPARK-23390](https://issues.apache.org/jira/browse/SPARK-23390) and relevant issues for resolving `FileBasedDataSourceSuite` (related to ORC dependency upgrade if I'm not missing here)

1) is pretty easy - cause they would be clean cherry-pick - but requires committer to drive efforts. So if someone who is committer interested on this, please take this up.
2) is a hard one though it incurs high chance of test failure. This seems to need some kind of decision so same as 1) I can't drive efforts. Please take this up if someone who is committer interested on this.

**Comment by maropu** [comment]
I'll cherry-pick SPARK-23491/23416 in brach-2.3.

**Comment by maropu** [comment]
@srowen @HeartSaVioR ok, done. Let's keep watching jenkins tests in branch-2.3.

**Comment by HeartSaVioR** [comment]
Thanks @maropu!

**Pull request closed without merging**
